File: tests/domain/task/taskStateMachine.test.ts
import { describe, expect, it } from "vitest";
import { isValidTaskTransition } from "../../../src/domain/task/taskStateMachine";

describe("taskStateMachine", () => {
  describe("isValidTaskTransition", () => {
    it("allows Pending → Ready_For_Execution", () => {
      expect(isValidTaskTransition("Pending", "Ready_For_Execution")).toBe(true);
    });

    it("allows Pending → Skipped", () => {
      expect(isValidTaskTransition("Pending", "Skipped")).toBe(true);
    });

    it("disallows Pending → Executing", () => {
      expect(isValidTaskTransition("Pending", "Executing")).toBe(false);
    });

    it("allows Ready_For_Execution → Executing", () => {
      expect(isValidTaskTransition("Ready_For_Execution", "Executing")).toBe(true);
    });

    it("allows Ready_For_Execution → Skipped", () => {
      expect(isValidTaskTransition("Ready_For_Execution", "Skipped")).toBe(true);
    });

    it("disallows Ready_For_Execution → Approved", () => {
      expect(isValidTaskTransition("Ready_For_Execution", "Approved")).toBe(false);
    });

    it("allows Executing → Awaiting_Review", () => {
      expect(isValidTaskTransition("Executing", "Awaiting_Review")).toBe(true);
    });

    it("allows Executing → Failed", () => {
      expect(isValidTaskTransition("Executing", "Failed")).toBe(true);
    });

    it("disallows Executing → Approved", () => {
      expect(isValidTaskTransition("Executing", "Approved")).toBe(false);
    });

    it("allows Awaiting_Review → Approved", () => {
      expect(isValidTaskTransition("Awaiting_Review", "Approved")).toBe(true);
    });

    it("allows Awaiting_Review → Rejected", () => {
      expect(isValidTaskTransition("Awaiting_Review", "Rejected")).toBe(true);
    });

    it("disallows Awaiting_Review → Executing", () => {
      expect(isValidTaskTransition("Awaiting_Review", "Executing")).toBe(false);
    });

    it("allows Approved → (no transitions, terminal state)", () => {
      expect(isValidTaskTransition("Approved", "Pending")).toBe(false);
      expect(isValidTaskTransition("Approved", "Ready_For_Execution")).toBe(false);
    });

    it("allows Rejected → Ready_For_Execution (rerun)", () => {
      expect(isValidTaskTransition("Rejected", "Ready_For_Execution")).toBe(true);
    });

    it("disallows Rejected → Approved", () => {
      expect(isValidTaskTransition("Rejected", "Approved")).toBe(false);
    });

    it("allows Skipped → (no transitions, terminal state)", () => {
      expect(isValidTaskTransition("Skipped", "Pending")).toBe(false);
      expect(isValidTaskTransition("Skipped", "Ready_For_Execution")).toBe(false);
    });

    it("allows Failed → Ready_For_Execution (rerun)", () => {
      expect(isValidTaskTransition("Failed", "Ready_For_Execution")).toBe(true);
    });

    it("disallows Failed → Approved", () => {
      expect(isValidTaskTransition("Failed", "Approved")).toBe(false);
    });
  });
});

File: src/test/java/com/wwa/deploymentagent/domain/task/TaskStateMachineTest.java
package com.wwa.deploymentagent.domain.task;

import com.wwa.deploymentagent.contracts.enums.TaskStatus;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

@DisplayName("TaskStateMachine")
class TaskStateMachineTest {

    @Test void pendingToReadyForExecution_allowed() {
        assertThat(TaskStateMachine.isValid(TaskStatus.Pending, TaskStatus.Ready_For_Execution)).isTrue();
    }

    @Test void pendingToSkipped_allowed() {
        assertThat(TaskStateMachine.isValid(TaskStatus.Pending, TaskStatus.Skipped)).isTrue();
    }

    @Test void pendingToExecuting_disallowed() {
        assertThat(TaskStateMachine.isValid(TaskStatus.Pending, TaskStatus.Executing)).isFalse();
    }

    @Test void readyToExecuting_allowed() {
        assertThat(TaskStateMachine.isValid(TaskStatus.Ready_For_Execution, TaskStatus.Executing)).isTrue();
    }

    @Test void readyToSkipped_allowed() {
        assertThat(TaskStateMachine.isValid(TaskStatus.Ready_For_Execution, TaskStatus.Skipped)).isTrue();
    }

    @Test void readyToApproved_disallowed() {
        assertThat(TaskStateMachine.isValid(TaskStatus.Ready_For_Execution, TaskStatus.Approved)).isFalse();
    }

    @Test void executingToAwaitingReview_allowed() {
        assertThat(TaskStateMachine.isValid(TaskStatus.Executing, TaskStatus.Awaiting_Review)).isTrue();
    }

    @Test void executingToFailed_allowed() {
        assertThat(TaskStateMachine.isValid(TaskStatus.Executing, TaskStatus.Failed)).isTrue();
    }

    @Test void executingToApproved_disallowed() {
        assertThat(TaskStateMachine.isValid(TaskStatus.Executing, TaskStatus.Approved)).isFalse();
    }

    @Test void awaitingReviewToApproved_allowed() {
        assertThat(TaskStateMachine.isValid(TaskStatus.Awaiting_Review, TaskStatus.Approved)).isTrue();
    }

    @Test void awaitingReviewToRejected_allowed() {
        assertThat(TaskStateMachine.isValid(TaskStatus.Awaiting_Review, TaskStatus.Rejected)).isTrue();
    }

    @Test void awaitingReviewToSkipped_allowed() {
        assertThat(TaskStateMachine.isValid(TaskStatus.Awaiting_Review, TaskStatus.Skipped)).isTrue();
    }

    @Test void awaitingReviewToExecuting_disallowed() {
        assertThat(TaskStateMachine.isValid(TaskStatus.Awaiting_Review, TaskStatus.Executing)).isFalse();
    }

    @Test void approvedIsTerminal() {
        assertThat(TaskStateMachine.isValid(TaskStatus.Approved, TaskStatus.Pending)).isFalse();
        assertThat(TaskStateMachine.isValid(TaskStatus.Approved, TaskStatus.Ready_For_Execution)).isFalse();
    }

    @Test void rejectedToReadyForExecution_allowed_rerun() {
        assertThat(TaskStateMachine.isValid(TaskStatus.Rejected, TaskStatus.Ready_For_Execution)).isTrue();
    }

    @Test void rejectedToApproved_disallowed() {
        assertThat(TaskStateMachine.isValid(TaskStatus.Rejected, TaskStatus.Approved)).isFalse();
    }

    @Test void skippedIsTerminal() {
        assertThat(TaskStateMachine.isValid(TaskStatus.Skipped, TaskStatus.Pending)).isFalse();
        assertThat(TaskStateMachine.isValid(TaskStatus.Skipped, TaskStatus.Ready_For_Execution)).isFalse();
    }

    @Test void failedToReadyForExecution_allowed_rerun() {
        assertThat(TaskStateMachine.isValid(TaskStatus.Failed, TaskStatus.Ready_For_Execution)).isTrue();
    }

    @Test void failedToApproved_disallowed() {
        assertThat(TaskStateMachine.isValid(TaskStatus.Failed, TaskStatus.Approved)).isFalse();
    }
}
